presto-hive/src/main/java/io/prestosql/plugin/hive/util/FSDataInputStreamTail.java

```diff
@@ -0,0 +1,113 @@
+{
+    public static final int MAX_SUPPORTED_PADDING_BYTES = 64;
+    private static final int MAXIMUM_READ_LENGTH = Integer.MAX_VALUE - (MAX_SUPPORTED_PADDING_BYTES + 1);
+
+    private final Slice tailSlice;
+    private final long fileSize;
+
+    private FSDataInputStreamTail(long fileSize, Slice tailSlice)
+    {
+        this.tailSlice = requireNonNull(tailSlice, "tailBuffer is null");
+        this.fileSize = fileSize;
+        checkArgument(fileSize >= 0, "fileSize must be >= 0, found: %s", fileSize);
+        checkArgument(tailSlice.length() <= fileSize, "length must be <= fileSize, found: %s > %s", tailSlice.length(), fileSize);
+    }
+
+    public static FSDataInputStreamTail readTail(String path, long paddedFileSize, FSDataInputStream inputStream, int length)
+            throws IOException
+    {
+        checkArgument(length >= 0 && length <= MAXIMUM_READ_LENGTH, "length too large, found: %s", length);
+        long readSize = min(paddedFileSize, (length + MAX_SUPPORTED_PADDING_BYTES));
+        long position = paddedFileSize - readSize;
+        // Actual read will be 1 byte larger to ensure we encounter an EOF where expected
+        byte[] buffer = new byte[toIntExact(readSize + 1)];
+        int bytesRead = 0;
+        long startPos = inputStream.getPos();
+        try {
+            inputStream.seek(position);
+            while (bytesRead < buffer.length) {
+                int n = inputStream.read(buffer, bytesRead, buffer.length - bytesRead);
+                if (n < 0) {
+                    break;
+                }
+                bytesRead += n;
+            }
+        }
+        finally {
+            inputStream.seek(startPos);
+        }
+        if (bytesRead > readSize) {
+            throw rejectInvalidFileSize(path, paddedFileSize);
+        }
+        return new FSDataInputStreamTail(position + bytesRead, Slices.wrappedBuffer(buffer, max(0, bytesRead - length), min(bytesRead, length)));
+    }
+
+    public static long readTailForFileSize(String path, long paddedFileSize, FSDataInputStream inputStream)
+            throws IOException
+    {
+        long position = max(paddedFileSize - MAX_SUPPORTED_PADDING_BYTES, 0);
+        long maxEOFAt = paddedFileSize + 1;
+        long startPos = inputStream.getPos();
+        try {
+            inputStream.seek(position);
+            int c;
+            while (position < maxEOFAt) {
+                c = inputStream.read();
+                if (c == -1) {
+                    return position;
+                }
+                position++;
+            }
+            throw rejectInvalidFileSize(path, paddedFileSize);
+        }
+        finally {
+            inputStream.seek(startPos);
+        }
+    }
+
+    private static PrestoException rejectInvalidFileSize(String path, long reportedSize)
+    {
+        throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Incorrect fileSize %s for file %s, end of stream not reached", reportedSize, path));
```

Comment:
```java
throw new PrestoException(HIVE_FILESYSTEM_ERROR, format("Incorrect file size (%s) for file (end of stream not reached): %s", reportedSize, path));

Reply:
This should throw `IOException` since it is used by `PrestoS3FileSystem`:
```java
throw new IOException(format("Incorrect file size (%s) for file (end of stream not reached): %s", reportedSize, path));